eth/stagedsync/stage_headers.go

```diff
@@ -135,13 +124,26 @@ func HeadersPOS(
 	cfg HeadersCfg,
 	initialCycle bool,
 	test bool, // Set to true in tests, allows the stage to fail rather than wait indefinitely
-	useExternalTx bool,
 ) error {
-	atomic.StoreUint32(cfg.waitingPosHeaders, 1)
+	// We need to have another write transaction for the miner in case we want to validate.
```

Comment:
Not sure I'm following. Why do you commit the transaction?

Reply:
we are closing it, in case we receive the request to mine. Mining stages uses a write transaction so we would stall when trying to propose new blocks

Reply:
Hmm. To gut feeling is that the transaction commitment here might be problematic. I don't understand the current interplay between waiting for PoS blocks and the mining. Probably need to pick up @AskAlexSharov brain for that.

Reply:
I do not think it is problematic, if any error occur we rollback them, so if the tx has not been killed and has been kept by another stage it should not give problems. if any stage does not rollback on error then the issue is not in the scope of stage headers. I can explain the interplay, mining stages uses a write transaction and block proposing happens only if we are waiting the beacon chain, thus when mining is requested the stagedsync must be in stage headers waiting for payloads. now if we did not close it prior waiting, the mining stage will never start unless we close the write transaction in stage headers. because you cannot have more than 2 write transactions on MDBX.

Reply:
Note also that external transactions are usually used for unit tests and not for production

Reply:
"external transactions are usually used for unit tests and not for production" - not true. 2 setups used in production: 
- if jump < 8K blocks - then whole staged sync step run inside 1 write transaction: means have only 1 fsync after new block processed, instead of fsync on each stage.
- if jump > 8K blocks (genesis sync, and long downtimes) - no external transaction started and stages can save partial progress as they want.

Reply:
About Mining and 2 write txs: 
Now, next code used to stop staged sync and release write tx - so, mining can start own write tx: 
```
case <-cfg.hd.SkipCycleHack:
			break Loop
```

Reply:
In my understanding: until "beacon validation done" we can't mine - because we don't know on which fork to mine (if beacon will say that block is valid - need mine on new block, if beacon say that block isn't valid - then on old block).

"waiting for beacon" - what is maximum time we can wait? As I understand it's synchronous call and we must just wait?

For example "waiting for p2p messages" is async - and implemented as `case <-cfg.hd.DeliveryNotify:`, can be interrupted by `case <-cfg.hd.SkipCycleHack:`. Then on next sync cycle all non-processed async messages will be handled.

Reply:
Yeah, we can probably use `SkipCycleHack` in `HeadersPOS` alongside `reverseDownloadCh` (listen on both). If something arrives to `SkipCycleHack`, then `HeadersPOS` should return so that the stage sync loop proceeds to the next stages, all the way to the mining.

Reply:
And when the Beacon node sends us a [engine_forkchoiceupdatedv1](https://github.com/ethereum/execution-apis/blob/v1.0.0-alpha.5/src/engine/specification.md#engine_forkchoiceupdatedv1) request with a payload, ethbackend should send a message to `SkipCycleHack` to interrupt the wait in the header stage.

Reply:
okay, so to answer all the points:
* Beacon chain calls are synchronous
* if external transactions are used in production then ending stage headers simply wont't be enough
* I noticed that there is an actual benefit of doing how i did it as of now, before we mined block every 3 seconds on a goroutine constantly, now we can do the same without having a goroutine constantly on and  only when requested.

Reply:
I think it makes more sense logically to do it like this, also tests are passing so it should not break any other component

Reply:
yes, must just wait until we timeout

Reply:
The thing is that Beacon mining ("payload build process") is supposed to be asynchronous. engine_forkchoiceUpdatedV1 with payload should only initiate the mining, not to wait for its completion. So IMHO more work is still required.

But this PR is too big already. In order to make progress, I suggest the following:
1. Leave only a rudimentary block build process in this PR, without any transactions (probably w/o correct state root either). Then we continue the work on proper PoS mining with transactions and everything in a subsequent PR.
2. Restore the `useExternalTx` logic in `HeaderPoS` and don't do this `tx.Commit` hack.